flask_dance/contrib/twitch.py

```diff
@@ -0,0 +1,111 @@
+        redirect_to (str, optional): If ``redirect_url`` is not defined, the name of the view to redirect to after the authentication dance is complete. Defaults to None.
+        login_url (str, optional): The URL path for the ``login`` view. Defaults to None.
+        base_url (str, optional): Base URL for the Twitch API. Defaults to "{Twitch API}/helix/".
+        authorization_url (str, optional): URL for the Twitch API login endpoint. Defaults to "{Twitch API}/oauth2/authorize".
+        token_url (str, optional): URL for Twitch token API. Defaults to "{Twitch API}/oauth2/token".
+        token_url_params (dict, optional): Options to send to the token URL. Defaults to {"include_client_id": True}.
+        authorized_url (str, optional): The URL path for the ``authorized`` view. Defaults to None.
+        session_class (class, optional):  The class to use for creating a Requests session. Defaults to None.
+        storage (class, optional): A token storage class, or an instance of a token storage class, to use for this blueprint. Defaults to None.
+
+    Raises:
+        MissingConfigException: If client_id or client_secret is missing when request is made.
+
+    Returns:
+        :class:`~flask_dance.consumer.OAuth2ConsumerBlueprint` A :ref:`blueprint <flask:blueprints>` to attach to your Flask app.
+    """
+    twitch_bp = OAuth2ConsumerBlueprint(
+        "twitch",
+        __name__,
+        client_id=client_id,
+        client_secret=client_secret,
+        scope=scope,
+        base_url=base_url,
+        authorization_url=authorization_url,
+        token_url=token_url,
+        token_url_params=token_url_params,
+        redirect_url=redirect_url,
+        redirect_to=redirect_to,
+        login_url=login_url,
+        authorized_url=authorized_url,
+        session_class=session_class,
+        storage=storage,
+    )
+    if twitch_bp.client_id is None:
+        twitch_bp.from_config["client_id"] = "TWITCH_OAUTH_CLIENT_ID"
+
+    if twitch_bp.client_secret is None:
+        twitch_bp.from_config["client_secret"] = "TWITCH_OAUTH_CLIENT_SECRET"
+
+    # TODO: The key won't auto renew. See https://github.com/singingwolfboy/flask-dance/issues/35 I think this will work but needs a test.
+    twitch_bp.auto_refresh_url = twitch_bp.token_url
+    twitch_bp.auto_refresh_kwargs = {
+        "client_id": twitch_bp.client_id,
+        "client_secret": twitch_bp.client_secret,
+    }
+
+    @twitch_bp.before_app_request
+    def set_applocal_session():
+
+        ctx = stack.top
+
+        if twitch_bp.client_id is None:
+            raise MissingConfigException(
+                f"client_id is required, currently is {twitch_bp.client_id}. Type: {type(twitch_bp.client_id)}"
+            )
+
+        if twitch_bp.client_secret is None:
+            raise MissingConfigException(
+                f"client_secret is required, currently is  is {twitch_bp.client_secret}. Type: {type(twitch_bp.client_secret)}"
+            )
```

Comment:
Is there a reason that you're doing this here? None of the other configurations behave this way -- which doesn't mean that this behavior is wrong, necessarily! I just want to understand why you're doing it. If we _are_ going to define a `MissingConfigException` class, it should probably be defined more broadly than just this specific configuration.

Reply:
I just thought it was a tad more descriptive than `TypeError` Its just something I started doing many moons ago after banging my head on a unhanded exception in a dependent library. Probably better to get included at a lower level. I can swap it out with `TypeError` easy enough. Its not so much that the type is wrong but rather missing. I took a glance at the python [error types ](https://docs.python.org/3/library/exceptions.html)and didn't see anything that really fit well. I'll take it in whatever direction you would like.

Reply:
Seems like a reasonable idea, but I think it's not related to adding the Twitch blueprint. As a result, it belongs in a separate pull request, so that we can discuss this idea separately -- and so that it can be applied to the whole codebase in a consistent way.

Please remove this `MissingConfigException` class from this pull request, and don't do this error checking here, to make it consistent with the other configurations. However, opening a new pull request that is focused only on adding a `MissingConfigException` class and applying it everywhere would be _very_ welcome!

Reply:
Will do! However I'm a little confused about your previous comment about throwing an exception. Are you suggesting that we just skip over that for now and come back to it in the separate PR? How should I handle the test for it if I'm not going to throw an error?  Sorry to be so needy in all this and thank you for being fantastically helpful!

>If Twitch won't allow you to use their API without this header, then I suggest throwing an exception if this value is not provided. You should have an automated test for that, too.

Reply:
Ah, that's a fair question. I asked you to do that when you were loading an environment variable in the `session_created()` function, and of course, environment variables could be defined or not defined. That's why I wanted you to raise an exception at that point: in case the environment variable was not defined. Also, this was an unusual case that was specific to the Discord configuration.

Now the situation has changed, and we're talking about something that is more broadly applicable to the entire codebase. In addition, we're no longer talking about a separate environment variable; we're talking about `twitch_bp.client_id` and `twitch_bp.client_secret`, which _must_ be defined; if they're not, the system will break in other, more interesting ways. (That's why I think the idea of raising a dedicated `MissingConfigException` is a good idea, but one that should be applied more globally.)

As a result, we don't need a test for what happens when these values are not defined -- at least, not until that other `MissingConfigException` pull request is created. We _do_ still need to test that the `Client-Id` header is applied to API requests to Twitch, but you've already done that, so I think we're good. 👍